Modify ApplicationSampler to user ApplicationStatus

cmcantalupo opened the pull request.

cmcantalupo commented on src/ApplicationSampler.hpp:
```diff
@@ -85,14 +87,8 @@ namespace geopm
-            /// @return Vector over Linux logical CPU of the region
-            ///         progress.
-            virtual std::vector<double> per_cpu_progress(void) const = 0;
+            virtual double cpu_hint_time(int cpu_idx, uint64_t hint) const = 0;
```
Doxygen missing

cmcantalupo commented on src/ApplicationSampler.hpp:
```diff
@@ -67,8 +67,10 @@ namespace geopm
+            ///        subsequent calls to the get_records()
+            ///        method. Also update cache of application
+            ///        status for use by hint and progress APIs.
+            virtual void update(const geopm_time_s &curr_time) = 0;
```
document curr_time

dianarg replied:
this is exposed for testing?

cmcantalupo replied:
Yes...might be nice to have a Clock singleton that we could mock for time values.

cmcantalupo reviewed (commented):
Needs some dox and some unit tests

dianarg commented on src/ApplicationSampler.cpp:
```diff
@@ -108,20 +111,41 @@ namespace geopm
 
-    ApplicationSamplerImp::ApplicationSamplerImp(const std::map<int, m_process_s> &process_map,
+const std::map<uint64_t, double> ApplicationSamplerImp::m_hint_time_init = {
+    {GEOPM_REGION_HINT_UNKNOWN, 0.0},
```
indentation

dianarg commented on src/Controller.cpp:
```diff
@@ -354,8 +354,10 @@ namespace geopm
         m_application_io->controller_ready();
-
         m_application_io->update(m_comm);
+        geopm_time_s curr_time;
+        geopm_time(&curr_time);
+        m_application_sampler.update(curr_time);
```
it would be nice to have a function returning the current time

dianarg commented on src/ApplicationSampler.cpp:
```diff
@@ -131,8 +155,18 @@ namespace geopm
+        if (!m_is_first_update) {
+            double time_delta = geopm_time_diff(&m_update_time, &curr_time);
+            for (int cpu_idx = 0; cpu_idx != m_num_cpu; ++cpu_idx) {
+                uint64_t hint = m_status->get_hint(cpu_idx);
```
status needs to guarantee hint is in range; no multiple hints

cmcantalupo committed:
Change ApplicationSampler to use AppliationStatus

- Update ApplicationSampler APIs to reflect ApplicationStatus usage
- Plumb the status usage by the application sampler
- Fix plumbing for per_cpu_process()
- Implement hint time and hint value.
- Add doxygen to ApplicationSampler.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg commented on test/ApplicationSamplerTest.cpp:
```diff
@@ -294,9 +302,9 @@ TEST_F(ApplicationSamplerTest, string_conversion)
-        .WillOnce(Return(expected));
+    std::vector<int> expected {42, 42};
+    EXPECT_CALL(*m_mock_status, get_process(_))
+        .WillRepeatedly(Return(42));
```
if you want to test different values here, you can chain together multiple ` .WillOnce(Return(..))`

dianarg commented on test/ApplicationSamplerTest.cpp:
```diff
@@ -359,3 +368,86 @@ TEST_F(ApplicationSamplerTest, short_regions)
+            .WillRepeatedly(Return(GEOPM_REGION_HINT_NETWORK));
+        m_app_sampler->update({{2, 0}});
+    }
+    double compute_time = m_app_sampler->cpu_hint_time(0, GEOPM_REGION_HINT_COMPUTE);
```
intermediate variables here are not really making the lines shorter

cmcantalupo committed:
Improve ApplicationSamplerTest tests

- Give different hint values on different CPUs

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
